lib/services/access_checker.go

```diff
@@ -0,0 +1,460 @@
+		return nil, trace.AccessDenied("failed to map roles for user with remote roles %v: %v", unmappedRoles, err)
+	}
+	if len(roles) == 0 {
+		return nil, trace.AccessDenied("no roles mapped for user with remote roles %v", unmappedRoles)
+	}
+	log.Debugf("Mapped remote roles %v to local roles %v and traits %v.",
+		unmappedRoles, roles, traits)
+
+	roleSet, err := FetchRoles(roles, access, traits)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	allowedResourceIDs, err := ExtractAllowedResourcesFromCert(cert)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &AccessInfo{
+		Roles:              roles,
+		Traits:             traits,
+		AllowedResourceIDs: allowedResourceIDs,
+		RoleSet:            roleSet,
+	}, nil
+}
+
+type RoleAndUserGetter interface {
+	RoleGetter
+	UserGetter
+}
+
+// AccessInfoFromLocalIdentity returns a new AccessInfo populated from the given
+// tlsca.Identity. Should only be used for cluster local users as roles will not
+// be mapped.
+func AccessInfoFromLocalIdentity(identity tlsca.Identity, access RoleAndUserGetter) (*AccessInfo, error) {
+	roles := identity.Groups
+	traits := identity.Traits
+	allowedResourceIDs := identity.AllowedResourceIDs
+
+	// Legacy certs are not encoded with roles or traits,
+	// so we fallback to the traits and roles in the backend.
+	// empty traits are a valid use case in standard certs,
+	// so we only check for whether roles are empty.
+	if len(identity.Groups) == 0 {
+		u, err := access.GetUser(identity.Username, false)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		log.Warnf("Failed to find roles or traits in x509 identity for %v. Fetching	"+
+			"from backend. If the identity provider allows username changes, this can "+
+			"potentially allow an attacker to change the role of the existing user.",
+			identity.Username)
+		roles = u.GetRoles()
+		traits = u.GetTraits()
+	}
+
+	roleSet, err := FetchRoles(roles, access, traits)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &AccessInfo{
+		Roles:              roles,
+		Traits:             traits,
+		AllowedResourceIDs: allowedResourceIDs,
+		RoleSet:            roleSet,
+	}, nil
+}
+
+// AccessInfoFromRemoteIdentity returns a new AccessInfo populated from the
+// given remote cluster user's tlsca.Identity. Remote roles will be mapped to
+// local roles based on the given roleMap.
+func AccessInfoFromRemoteIdentity(identity tlsca.Identity, access RoleGetter, roleMap types.RoleMap) (*AccessInfo, error) {
+	// Set internal traits for the remote user. This allows Teleport to work by
+	// passing exact logins, Kubernetes users/groups and database users/names
+	// to the remote cluster.
+	traits := map[string][]string{
+		teleport.TraitLogins:     identity.Principals,
+		teleport.TraitKubeGroups: identity.KubernetesGroups,
+		teleport.TraitKubeUsers:  identity.KubernetesUsers,
+		teleport.TraitDBNames:    identity.DatabaseNames,
+		teleport.TraitDBUsers:    identity.DatabaseUsers,
```

Comment:
What about `TraitWindowsLogins`?

Reply:
Gotcha. Probably because we don't support trusted clusters with desktop access yet.